### DISCUSSION TITLE
ModDB download helper

### DISCUSSION THREAD

**Discussion by karthanistyr** [lutris/lutris#3872] [Ideas]
Hi everyone,
I wanted to start a discussion to take the temperature regarding a possible ModDB download helper being built within lutris?

### **Why**
ModDB is a large repository of mods for many games, and sometimes outright standalone games. Installer scripts would benefit being able to reliably pull resources from ModDB, availing many customisations and new games to be installed in a single click in Lutris.

### **Current behaviour**
ModDB direct links are temporary and rotate periodically. Installers that incorporate direct links to ModDB mirrors will stop functioning after a few days (hours?).

### **Proposal**
Automated downloads from ModDB require a 2-step process: call a first, stable, permanent URL, and scrape the result of that request for a valid mirror link, which points to the actual file we want to download.

Here's what I posted on discord ([link to discord](https://discord.com/channels/512538904872747018/799769240471207996/914880798409240667)):

> would there be appetite for a patch in lutris for supporting downloading a moddb file from the "permalink" ? Direct mirror links in moddb rotate periodically, so in order to get an active one, we would need to call `https://www.moddb.com/downloads/start/<file_id>` (this url is permanent) and scrape the returned html for the actual file URL, e.g. `https://www.moddb.com/downloads/mirror/<file_id>/<server_id>/<some_hash>`
> 
> This would make installers pulling files from moddb far easier to write and allow for fully automated installs (for now the only reliable way to involve moddb is to ask the user to download the file manually and use a "N/A:" prefix in the files section of the installer)

There are example implementations of such process, including a script suggested by discord user **Keenan**: 
* https://aur.archlinux.org/cgit/aur.git/tree/moddb-downloader.sh?h=realrtcw (via Keenan)
* https://github.com/EpicMorg/modDB_downloader
* https://github.com/ClementJ18/moddb

### Benefits: 
no-brainer, built-in support for linking and downloading moddb resources within installer scripts

### **Risks**: 
we would not rely on a stable API to provide the mirror links, therefore may be slightly more prone to breakage if the ModDB website changes. However, from memory, the 2-step download process has been as it is now for multiple years.

There are two possible outcomes in case of a significant change in the ModDB website: either the primary links (`/download/start/<id>`) are removed, and this isn't different from a simple broken link anywhere else, or the current 2-step process changes and we would retain some possibility of control to repair, adjust or otherwise correct the built-in helper.

Please indicate if you would think this valuable for Lutris.

**Comment by legluondunet** [comment]
I don't think it's such  a good idea. Moddb is a good site where you can download very interesting mods, it's a reference in the gaming domain. All the traffic it gerenates with visits and downloads cost money. So the website proprietary need people go to their site and see pubs. Less visits, less money and this oculd fragilize the site...Is it a good way to thanks Moddb website owner? I don't think so.

**Comment by AlexanderRavenheart** [comment]
I think it is a great idea, so that we don't have to rely on user maintained repos, google drives, dropboxes or other sources of hosting patches, mods and other files used in installers or requesting the user to go on site X, Y, Z to download files as the whole point of Lutris is to streamline the process as much as possible.

As far as that whole idea of users going to their website and supporting them with ads? Pretty sure they get literal pennies from that because most users will have some sort of ad blocker to get rid of all of those.

Ads as an income source for websites should have died a long time ago, as it just infects good website designs and ruins the user experience regardless of which types of ads are used. Nobody wants to see an ad when they want to just download X mod to play their game with Y feature.

Sponsorships, subscriptions and donations are how you make a website be sustainable in the long-term.

**Comment by legluondunet** [comment]
@AlexanderRavenheart it's not to you or me to decide, but to moddb website owner if we can download directly files they host.
We need their agreement in first place.

**Comment by intenscia** [comment]
Hey all - Scott here creator of ModDB.com. Firstly - love Lutris and the amazing service you provide the community.

Our aim is to always provide a safe place for mod creators to collaborate and share their mods, knowing they will be showcased and available for many years to come (next year is our 20th anniversary).

@AlexanderRavenheart while you are right that ads can infect good website design, subscriptions also can since sites that push subscriptions will often cripple functionality to encourage adoption. In our case we rely on ads, and we've repeatedly turned down ads that overlay, interefere or play sounds - because we'd rather provide a great user experience than chaise extra $ and annoy users in the process. We also don't penalize users that adblock, we understand your reasons since not ever site shares the same morals.

We are happy to support Lutris in attempting to make mods more accessible (it's our goal too), so if you want to implement a system like this you have our support. Our only ask is that you always ensure clear links directly to the content and ModDB are provided, creators are properly credited, and our servers / bandwidth respected by your implementation. We reserve the right to change this, as our priority is always to the creators and community visiting via ModDB, and ensuring we can continue to archive their content in perpetuity.

Thanks for thinking of us @legluondunet and @karthanistyr

**Reply by karthanistyr** [reply]
what a twist, thanks a lot for taking the time to weigh in! Love moddb and what you are doing for the modding community.

@legluondunet does this alleviate your concerns ?

**Reply by legluondunet** [reply]
yes of course

**Comment by strycore** [comment]
I'm in favor of mirroring the files and providing the bandwidth for the files that are used in Lutris installers. I would prefer that over working around ModDB's system to automate downloads from their servers. 
The download helper can still be used by our backend to create file mirrors. 

I'm finding out about the moddb API ( https://moddb.readthedocs.io/en/latest/ ) and I think we could add that as an integration to our website just like we did with IGDB. That kind of integration seems important if we want to be able to link back to the creator's page. 

@intenscia what do you think? Would mirroring the files we have installers for make the most sense here?

**Reply by intenscia** [reply]
That API is unofficial, but ok to use. Caching the files is akin to copying and bypassing the site, so we don't permit it as it's unfair on the creators who may want to control and restrict where or how their content is distributed.

**Reply by strycore** [reply]
got it, we'll see what kind of integration we can work out.
